pkg/kubelet/kubelet.go

```diff
@@ -2316,29 +2316,68 @@ func GetPhase(spec *api.PodSpec, info []api.ContainerStatus) api.PodPhase {
 	}
 }
 
+func readyPodCondition(isPodReady bool, reason, message string, existingStatus *api.PodStatus) []api.PodCondition {
+	currentTime := util.Now()
+	condition := api.PodCondition{
+		Type: api.PodReady,
+	}
+	if isPodReady {
+		condition.Status = api.ConditionTrue
+	} else {
+		condition.Status = api.ConditionFalse
+	}
+	condition.LastProbeTime = currentTime
+	condition.Reason = reason
+	condition.Message = message
+	if existingStatus != nil {
+		if api.IsPodReadyConditionTrue(*existingStatus) != isPodReady {
+			// condition has transitioned, update transition time.
+			condition.LastTransitionTime = currentTime
+		} else {
+			// retain the existing transition time.
+			existingCondition := api.GetPodReadyCondition(*existingStatus)
+			if existingCondition != nil {
+				condition.LastTransitionTime = existingCondition.LastTransitionTime
+			}
+		}
+	}
+	return []api.PodCondition{condition}
+}
+
 // getPodReadyCondition returns ready condition if all containers in a pod are ready, else it returns an unready condition.
-func getPodReadyCondition(spec *api.PodSpec, statuses []api.ContainerStatus) []api.PodCondition {
-	ready := []api.PodCondition{{
-		Type:   api.PodReady,
-		Status: api.ConditionTrue,
-	}}
-	unready := []api.PodCondition{{
-		Type:   api.PodReady,
-		Status: api.ConditionFalse,
-	}}
-	if statuses == nil {
-		return unready
+func getPodReadyCondition(spec *api.PodSpec, containerStatuses []api.ContainerStatus, existingStatus *api.PodStatus) []api.PodCondition {
+	// Find if all containers are ready or not.
+	if containerStatuses == nil {
+		return readyPodCondition(false, "UnknownContainerStatuses", "", existingStatus)
 	}
+	containersWithUnknownStatus := []string{}
+	containersWithUnreadyStatus := []string{}
 	for _, container := range spec.Containers {
-		if containerStatus, ok := api.GetContainerStatus(statuses, container.Name); ok {
+		if containerStatus, ok := api.GetContainerStatus(containerStatuses, container.Name); ok {
 			if !containerStatus.Ready {
-				return unready
+				containersWithUnreadyStatus = append(containersWithUnreadyStatus, container.Name)
 			}
 		} else {
-			return unready
+			containersWithUnknownStatus = append(containersWithUnknownStatus, container.Name)
 		}
 	}
-	return ready
+	unreadyMessage := ""
+	if len(containersWithUnknownStatus) > 0 {
+		unreadyMessage = fmt.Sprintf("Containers with unknown status: %s", containersWithUnknownStatus)
+	}
+	if len(containersWithUnreadyStatus) > 0 {
+		if unreadyMessage == "" {
+			unreadyMessage = fmt.Sprintf("Containers with unready status: %s", containersWithUnreadyStatus)
```

Comment:
Just a nit (and you are free to ignore it). 

If capitalization is not important here, I'd suggest creating a []string, append it, and join them in the end to create the unready message.